src/cache/cache.go

```diff
@@ -16,9 +16,18 @@ type Cache interface {
 	// Returns the value and a boolean indicating if the key was found.
 	// In case the duration expired, the function returns false.
 	Get(key string) (string, bool)
+	// Gets the value for a given key and tag.
+	// Returns the value and a boolean indicating if the key was found.
+	// In case the duration is expired or the tag does not match the cached tag,
+	// the function returns false.
+	GetTagged(key, tag string) (string, bool)
```

Comment:
I don't see why this needs to be part of this interface, a `SegmentWriter `can implement another interface for this to work and we can use the key in combination with a tag as the actual key value, right? It doesn't need to change the `Cache` interface.

Reply:
I suppose that could work, though we wouldn't even need a new interface in that case: just return the current `CacheKey()` + the file content from `CacheKey()`, but wouldn't that mean your current session's cache file keeps growing with each branch change? I'll try it and see.

Alternatively, the value stored could be a wrapper such that the key doesn't change but we store and compare the tag along with the serialized `writer`.

Reply:
Yeah, just changing the key will work. I wasn't considering that the cache is rewritten each time so there's no risk of growth (apart from lots of cache files being created normally, but a cronjob could clean that up).

I'll add an interface for segments to implement who want to provide an extra tag to their key and we can keep the `Cache` interface as it is in main.

Reply:
You can actually reuse [this](https://github.com/JanDeDobbeleer/oh-my-posh/blob/a9a8dcd4f2c120deed189a865ce4c55de779a99e/src/config/segment.go#L274) logic. It's implemented [here](https://github.com/JanDeDobbeleer/oh-my-posh/blob/a9a8dcd4f2c120deed189a865ce4c55de779a99e/src/segments/git.go#L214) for Git for example, we can extend that using the logic you have in mind.

Reply:
I have something working, but there's a lot of duplication in `git.go` e.g., setting `g.realDir`, `g.Ref`, etc., I'm trying to clean up with introducing any regressions.

Reply:
Not sure it's worth it. There are several "entry points" into the `git.go` file that need different information at different times. Most access is simply checking file paths, which should be pretty quick. I need to get file content, which some of the similar code paths don't need to do. On top of that, some fields e.g., `g.Ref` may get set later to something more...accurate...once `git` command lines are run. So maybe it's safer like this?

Reply:
@heaths with this you mean the current code changes? I know the git segment is rather heavy. One of those parts in the codebase that grew disproportionately.

Reply:
Sorry. I meant that after following several code paths from `segment.go` into `git.go` (and `scm.go`), I wasn't sure it was worth refactoring - not without a major overhaul anyway. While there were several code paths that do at least basic checks for whether we're in a main or linked worktree, not all of them need to read the `HEAD` file content. Checking file paths repeatedly isn't so expensive, I'd think, but adding file content reads unnecessarily is likely worse.

Perhaps worth a separate change to consolidate incremental updates to various members e.g., `g.rootDir`, `g.Ref`, and a few others could be set early e.g., from `shouldDisplay()`, which gets called early in most code paths into `git.go`?

Reply:
@heaths reading the contents isn't that crazy expensive TBH. You can see it here:

```
[TRACE] 20:46:02.503 scm.go:FileContents(D:\oh-my-posh\.git/HEAD) - 0s
```